src/Microsoft.Data.Sqlite/Interop/NativeMethods.cs

```diff
@@ -243,15 +255,30 @@ public static int sqlite3_prepare_v2(Sqlite3Handle db, string zSql, out Sqlite3S
         }
 
         [DllImport("sqlite3", CallingConvention = CallingConvention.Cdecl)]
+        public static extern int sqlite3_reset(Sqlite3StmtHandle stmt);
+
+        [DllImport("sqlite3", CallingConvention = CallingConvention.Cdecl)]
         public static extern int sqlite3_step(Sqlite3StmtHandle stmt);
 
+        public static int sqlite3_step_blocking(Sqlite3Handle db, Sqlite3StmtHandle stmt, int milliseconds)
+        {
+            var timer = new Stopwatch();
+            int rc;
+            timer.Start();
+            while (SQLITE_LOCKED == (rc = sqlite3_step(stmt)))
+            {
+                if (timer.ElapsedMilliseconds >= milliseconds)
+                {
+                    return rc;
+                }
+                sqlite3_reset(stmt);
+                Thread.Sleep(150);
```

Comment:
Can of worms: We could now have an async code path that calls `Task.Delay()` here instead. If nothing else, we should create an issue.